### PR TITLE
Introduce error code for errors

### LINKED ISSUE #919: Raising an exception in a POST request returns with a 200 response
The following request:

```
curl -XPOST -d '{"name": "e", "type": "ManageIQ::Providers::EmbeddedAnsible::AutomationManager::VaultCredential", "vault_password": "x"}' "http://admin:smartvm@localhost:3000/api/authentications" | jq
```

Returns a `200` response:

```
{
  "results": [
    {
      "success": false,
      "message": "must supply a manager resource"
    }
  ]
}
```

Is this correct? Because the UI codebase expects HTTP errors if the request failed :confused:  

@miq-bot add_label bug

### PR DESCRIPTION
**Pull request by kbrock**
Fixes #919 

We are currently returning an ok (200) http status from all api requests, even the ones that fail.
This makes routing difficult in the ui.

It is not so cut and dry because some requests take multiple records and end up with multiple statuses, some failing and some succeeding.

### Before 

We returned an ok for all create and update requests, even when they failed.

### After

Now if there is one resource, including a collection of one resource, and it fails creating or updating, it will return a failed status code (4xx)

The status of `406` (not processable) is the best status available to us.

### CODE REVIEW

File: app/controllers/api/base_controller/renderer.rb
Comment (chessbyte) [inline app/controllers/api/base_controller/renderer.rb:629]: `: not_acceptables` is the plural a typo?

Comment (kbrock) [reply app/controllers/api/base_controller/renderer.rb:629]: totally - thanks

**Comment by Fryguy** [comment]
I don't understand using 406 for singles that fail...feels weird.   406 is like when the client does something like `Accept: text/xml` content and the server doesn't provide xml content, so it replies not acceptable.

**Comment by Fryguy** [comment]
The only 400-level codes that makes sense to me, in this case, are 

- 422 - Unprocessable Entity https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422
- 400 - Bad Request

422 was designed for WebDAV, which is why I'm leaning towards just a normal 400 error.

**Review by Fryguy — CHANGES_REQUESTED** [review]
Code wise this looks good to me, though I'd like another set of eyes.  My request changes is mainly around the choice of error code, as I think 406 is incorrect.

File: app/controllers/api/base_controller/renderer.rb
Comment (Fryguy) [inline app/controllers/api/base_controller/renderer.rb:621]: Can you comment on this line...I understand it only because I understand the issue, but a casual reader won't understand why a non-hash is atuomatically :ok

Comment (kbrock) [reply app/controllers/api/base_controller/renderer.rb:635]: we sure have a lot of ways we can render objects. I hope the documentation will serve well for people (including me) in the future

**Comment by abellotti** [comment]
Maybe a quick sanity check with https://github.com/ManageIQ/manageiq-api-client's usage just to make sure we're not breaking central admin.

**Comment by kbrock** [comment]
> Maybe a quick sanity check with https://github.com/ManageIQ/manageiq-api-client's usage just to make sure we're not breaking central admin.

I do see a response code of 400 triggering a special response in [`connection.rb`](https://github.com/ManageIQ/manageiq-api-client/blob/master/lib/manageiq/api/client/connection.rb#L109) 

I'm having trouble knowing if we do or do not want that exception thrown. Since we weren't before, then I'm assuming we would not like that now. Then again, before it is possible that we had to add special logic to handle his and that will be fixed.

do you know of any other cases in our code that may also react negatively to this code?

TL;DR - nice find. I'm stumped where to go from here

**Review by Fryguy — APPROVED** [review]
Code-wise seems good to me.  @abellotti Please also review.  As @abellotti we should just make sure that the manageiq-api-client responds accordingly.  If it's a bad request, I think I'd expect the client to raise an Exception.  Since we use the client in core in order to do the central admin stuff (and also manageiq-automation_engine I believe), I'm assuming there will be tests for it, so perhaps a cross-repo test is enough?

File: lib/api/request_adapter.rb
Comment (abellotti) [inline lib/api/request_adapter.rb:108]: may be easier to understand `json_body.kind_of?(Hash) && json_body.key?("resources")`

File: spec/requests/authentications_spec.rb
Comment (abellotti) [inline spec/requests/authentications_spec.rb:266]: why is this needed in this PR ?

File: lib/api/request_adapter.rb
Comment (Fryguy) [reply lib/api/request_adapter.rb:108]: @abellotti That's not the same thing...

`!A || B` -> `A && !B`

Comment (abellotti) [reply lib/api/request_adapter.rb:108]: But it is what we want though, a Hash that has a key resources.

Comment (Fryguy) [reply lib/api/request_adapter.rb:108]: I was assuming the `!json_body.kind_of?(Hash)` was referring to the Array notation as listed in the comment.  That is I interpreted it as:

```
!json_body.kind_of?(Hash) || json_body.key?("resources")

Array || Hash-with-resources
```

Comment (kbrock) [reply lib/api/request_adapter.rb:108]: |json_body|bulk?|
|---|---|
|hash with :resources|true|
|hash without :resources|false|
|array of hashes|true|
|nil| don't care|
|some other class|don't care


I will make the successes more explicit rather than the vague `not`

Comment (abellotti) [reply lib/api/request_adapter.rb:108]: I may be missing something but I don't recall the second signature as bulk. @kbrock which API endpoint implements this ?

I know of the PATCH /api/:collection/:id with the similar signature as per https://manageiq.org/docs/reference/latest/api.overviiew/http.html example of PATCH /api/services/42, Array of attribute update specs, so really a single resource update, not bulk.

File: spec/requests/authentications_spec.rb
Comment (kbrock) [reply spec/requests/authentications_spec.rb:266]: deleted it... remembered why it is there.

I wanted to ensure that if we have a successful and a failed creation of an authentication (it requires a manager_resource to create the record) that it would still return a 200.

**Comment by kbrock** [comment]
@Fryguy 

The api client simply raises for any response code >= 400 (with a special case for 404). So this will be handled.

From what I can tell, the api client tests do not use the api to test so a cross repo would probably not buy us anything.

**Comment by kbrock** [comment]
Kicking

**Comment by kbrock** [comment]
pushed a rebase (no changes)

I was reproducing locally and after this rebase, it fixed it locally.

**Review by gtanzillo — APPROVED** [review]
This looks good to me. I'd like to get a final 👍  from @abellotti before merging if it can wait another day

**Comment by Fryguy** [comment]
@gtanzillo I think this thread is still unresolved - https://github.com/ManageIQ/manageiq-api/pull/961#discussion_r529727436

**Comment by kbrock** [comment]
update:

i had thought that passing an array of a hash in the specs meant that it supported an array of 2 hashes. this was wrong. The `Hash(param)` where `param=[{}]` voided it out.
I had thought that meant that just passing a hash always worked, but this turned out to be wrong as well. While all but the chargeback example worked, chargeback was throwing odd errors - and not changing this behavior seemed best. especially since it made no sense to me.

**Review by abellotti — APPROVED** [review]

**Comment by Fryguy** [comment]
@miq-bot cross-repo-test manageiq-ui-classic,manageiq

**Review by Fryguy — APPROVED** [review]
LGTM - just waiting on cross repo test.

**Pull request merged**
